### ISSUE TITLE
[ddc/dart2js] Compilers assume constructors that are external, only contain named args, and are extension type members are object literal constructors

### ISSUE THREAD

**Issue by srujzs** [dart-lang/sdk#54968]
This is tracking cleanup of an assumption we have in both compilers.

With interop extension types, we're allowed to write code like this:

```dart
extension type Options._(JSObject _) implements JSObject {
  external Options({int a, int b});
}
```

where the `Options` constructor is an object literal constructor. This type of constructor is special because it's the only interop member that is lowered only in the backends in order to support a more performant syntax. Since we avoid using the `JS<>` foreign helper in our interop transformers, they do not have a way to lower something like `Options({a: 1, b: 1})` to `{a: 1, b: 1}`. Instead, they would have to do some combination of creating a new object and using property setters to do this. So, we do this instead in the backends.

However, determining that this constructor *is* an object literal constructor can be expensive. It requires indexing extension types and doing a reverse lookup to determine that this is an interop member and a constructor. This indexing is already done in the interop transformer, but that's not where we do the lowering. Re-indexing just for this one case is likely expensive. So, we instead make the assumption that any extension type member that is `external` and only contain named args are object literal constructors. JS interop has checks to ensure that the only JS interop `external` members that have named args are object literal constructors, but naturally `external` can be used for other purposes, so this assumption may prove false sometime in the future.

Options are:

1. Avoid the need for indexing by having the AST providing members on extension type members to access their extension type.
2. Provide some mechanism to the interop transformer to create an object literal.
3. Re-index extension types in the backends.

**Comment by fishythefish** [comment]
For option 1, does [`Member.enclosingExtensionTypeDeclaration`](https://github.com/dart-lang/sdk/blob/ac39be3eed71a6bb97d86780d729e5d9bbfee9c2/pkg/kernel/lib/ast.dart#L2122) work?

Also, for context, why do we avoid using `JS` in our transformers?

**Comment by srujzs** [comment]
> For option 1, does [Member.enclosingExtensionTypeDeclaration](https://github.com/dart-lang/sdk/blob/ac39be3eed71a6bb97d86780d729e5d9bbfee9c2/pkg/kernel/lib/ast.dart#L2122) work?

Ooh, I didn't know this was added. We could probably optimize some of our existing indexing to use this (maybe). As for whether we could use this in the backends to avoid the more expensive lookup, it certainly helps, but it still doesn't quite get all the way there as we determine whether something is a JS interop type based on whether it wraps a JS type recursively. We can't quite use `extensionTypeErasure` for that as JS types are extension types themselves, so there's still a bit of cost there.

> Also, for context, why do we avoid using `JS` in our transformers?

It's been a while since we've had this discussion, so I don't remember much beyond us wanting to limit the scope of where it's used. Maybe @rakudrama remembers why.

**Comment by fishythefish** [comment]
We've been talking about using [`Component.metadata`](https://github.com/dart-lang/sdk/blob/ac39be3eed71a6bb97d86780d729e5d9bbfee9c2/pkg/kernel/lib/ast.dart#L14701) in dart2js for various reasons. Could we apply it here? The initial interop transformer can add a metadata mapping to tag relevant nodes, and we can simply read it later in the backends rather than reindexing.

**Comment by srujzs** [comment]
Yes. We know the node in question in the interop transformer so tagging it and having the backend read it later would definitely work.
